src/Servers/Kestrel/Core/src/Internal/WebTransport/WebTransportSession.cs

```diff
@@ -0,0 +1,191 @@
+    }
+
+    internal void OnClientConnectionClosed()
+    {
+        if (_isClosing)
+        {
+            return;
+        }
+
+        _isClosing = true;
+
+        _connectionClosedRegistration.Dispose();
+
+        lock (_openStreams)
+        {
+            foreach (var stream in _openStreams)
+            {
+                stream.Value.DisposeAsync().AsTask().GetAwaiter().GetResult();
+            }
+
+            _openStreams.Clear();
+        }
+
+        _pendingStreams.Writer.Complete();
+    }
+
+    internal void Abort(ConnectionAbortedException exception, Http3ErrorCode error)
+    {
+        if (_isClosing)
+        {
+            return;
+        }
+
+        _isClosing = true;
+
+        _connectionClosedRegistration.Dispose();
+
+        lock (_openStreams)
+        {
+            _connectStream.Abort(exception, error);
+            foreach (var stream in _openStreams)
+            {
+                if (exception.InnerException is not null)
+                {
+                    stream.Value.Abort(new ConnectionAbortedException(exception.Message, exception.InnerException));
+                }
+                else
+                {
+                    stream.Value.Abort(new ConnectionAbortedException(exception.Message));
+                }
+            }
+            _openStreams.Clear();
+        }
+
+        _pendingStreams.Writer.Complete();
+    }
+
+    public async ValueTask<ConnectionContext?> OpenUnidirectionalStreamAsync(CancellationToken cancellationToken)
+    {
+        if (_isClosing)
+        {
+            return null;
+        }
+        // create the stream
+        var features = new FeatureCollection();
+        features.Set(_outputStreamDirectionFeature);
+        var connectionContext = await _connection._multiplexedContext.ConnectAsync(features, cancellationToken);
+        var streamContext = _connection.CreateHttpStreamContext(connectionContext);
+        var stream = new WebTransportStream(streamContext, WebTransportStreamType.Output);
+
+        var success = _openStreams.TryAdd(stream.StreamId, stream);
+        Debug.Assert(success);
+
+        // send the stream header
+        // https://ietf-wg-webtrans.github.io/draft-ietf-webtrans-http3/draft-ietf-webtrans-http3.html#name-unidirectional-streams
+        await stream.Transport.Output.WriteAsync(OutputStreamHeader, cancellationToken);
+
+        return stream;
+    }
+
+    internal void AddStream(WebTransportStream stream)
+    {
+        if (_isClosing)
+        {
+            return;
+        }
+
+        var addedToOpenStreams = _openStreams.TryAdd(stream.StreamId, stream);
+
+        if (!addedToOpenStreams || !_pendingStreams.Writer.TryWrite(stream))
+        {
+            if (addedToOpenStreams)
+            {
+                _openStreams.Remove(stream.StreamId, out _);
+            }
+
+            stream.Abort();
+            throw new Exception(CoreStrings.WebTransportFailedToAddStreamToPendingQueue);
```

Comment:
How is this exception handled?

Reply:
The exception is passed back up to the `Http3Connection` class where it is caught. The abort is here and not there because I don't have access to the stream object in the catch block.

Reply:
Actually, since we are already talking about it, do we need logs for this? Currently it just throws, catches and moves on.

Reply:
There should be logs where exceptions are caught and not re-thrown. We could also consider logging here and not throwing if the caller isn't going to behave any differently with/out an exception.

Reply:
The stream.Abort() has logging inside of it. So, we can log that way. Updating.